manager-bundle/src/ContaoManager/Plugin.php

```diff
@@ -464,6 +484,42 @@ private function addDefaultMailer(array $extensionConfigs, ContainerBuilder $con
         return $extensionConfigs;
     }
 
+    /**
+     * @return array{0: string|null, 1: array<string, mixed>}
+     */
+    private function parseDbalDriverAndOptions(array $extensionConfigs, ContainerBuilder $container): array
+    {
+        $driver = null;
+        $url = null;
+        $options = [];
+
+        foreach ($extensionConfigs as $config) {
+            if (null !== ($driverConfig = $config['dbal']['connections']['default']['driver'] ?? null)) {
+                $driver = $driverConfig;
+            }
+
+            if (null !== ($urlConfig = $config['dbal']['connections']['default']['url'] ?? null)) {
+                $url = $container->resolveEnvPlaceholders($urlConfig, true);
+            }
+
+            if (null !== ($optionsConfig = $config['dbal']['connections']['default']['options'] ?? null)) {
+                $options = array_replace($options, $optionsConfig);
```

Comment:
Can we optimize this?

<img width="757" alt="" src="https://user-images.githubusercontent.com/1192057/152526185-dc6ac939-e661-4eaf-9e53-29f282fc2f28.png">

Reply:
```suggestion
                $options = $optionsConfig + $options;
```
Would this one be more performant? (And does it work the same? ☺️)

Reply:
For `array_merge()`, this can usually be fixed like this: https://3v4l.org/Gl8Op

```php
$a = [1];
$n = [[2, 3, 4], [5, 6, 7]];

// Multiple calls to array_merge
foreach ($n as $m) {
    $a = array_merge($a, $m);
}

$a = [1];
$b = [$a];

foreach ($n as $m) {
    $b[] = $m;
}

// Single call to array_merge
$a = array_merge(...$b);
```

Maybe the same can be done for `array_replace()`?

Reply:
There is literally no performance loss by using `array_replace` in this case. Feel free to change it, but this is IMHO nothing more than code style.

To the different options: Array merging does not work, because keys are reassigned, spread doesn't work with keys except for recent PHP versions, + means the same, but most of the time people argue that it could be confused with an arithmetic operation.

Reply:
> There is literally no performance loss by using `array_replace` in this case.

Why not in this case?

Reply:
There a way too few iterations.

Reply:
At this point, the discussion is becoming pointless.

<img width="479" alt="" src="https://user-images.githubusercontent.com/1192057/152610927-3caa1efa-2b1d-474a-bc9c-2f78cfc7bf8d.png">

Any discussion beyond the scope of this decision process is most likely just a waste of time.

Reply:
The inspector cannot know about iterations and the impact. Blindly trusting inspections is not helpful at all. They warn you about possible pitfalls, but you, as a developer, must ultimately be able to judge if this is relevant in your case or not.

If you _really_ believe this must be implemented differently, please go ahead. As I said, for me it's code style at this point.

Reply:
This is exactly how technical debt is created!

On your system, there are only two or three iterations, so you think "this is not relevant" and ignore it. But out in the wild, there might be systems with fifty or sixty iterations or very huge arrays and suddenly tools like Blackfire identify this exact function call as a bottleneck.

I don‘t understand why a developer would not want to prevent such a problem in the first place? Especially if they have already been made aware of it.

We usually care about defensive programming but in this case it is OK not to do it? IMHO, this is not "judgement" but ignorance. No hard feelings.

Reply:
```suggestion
                $options[] = $optionsConfig;
```

and at the end:

```php
$options = array_replace([], ...$options);
```

should work IMO.

But I would not say it’s ignorance as `$config['dbal']['connections']['default']['options']` wil never be set in fifty configs and will not be huge itself. So I would be happy with both versions 🥰